.kokoro/build.sh

```diff
@@ -0,0 +1,237 @@
+        --region=${DATAPROC_CLUSTER_REGION} \
+        --jars=${BIGTABLE_SPARK_JAR} \
+        ${TEST_SCRIPT} \
+        --py-files=${BASE_SCRIPT} \
+        -- \
+        --bigtableProjectId=${BIGTABLE_PROJECT_ID} \
+        --bigtableInstanceId=${BIGTABLE_INSTANCE_ID} \
+        --bigtableTableId=${TABLE_ID}
+    exit_code=$?
+    delete_table ${BIGTABLE_PROJECT_ID} ${BIGTABLE_INSTANCE_ID} ${TABLE_ID}
+    return $exit_code
+}
+
+run_fuzz_tests() {
+    SPARK_VERSION=$1
+    echo "***Running Spark-Bigtable fuzz tests for Spark ${SPARK_VERSION}.***"
+    TABLE_ID=$(create_table_id "fuzz")
+    create_table_with_random_splits \
+      "${BIGTABLE_PROJECT_ID}" "${BIGTABLE_INSTANCE_ID}" "$TABLE_ID"
+    BIGTABLE_SPARK_IT_MODULE="spark-bigtable_2.12-it"
+    mvn -pl ${BIGTABLE_SPARK_IT_MODULE} \
+        failsafe:integration-test failsafe:verify \
+        -B -ntp -Dclirr.skip=true -Denforcer.skip=true -Dcheckstyle.skip \
+        -Dspark.version="${SPARK_VERSION}" \
+        -DbigtableProjectId="${BIGTABLE_PROJECT_ID}" \
+        -DbigtableInstanceId="${BIGTABLE_INSTANCE_ID}" \
+        -DbigtableTableId="${TABLE_ID}" \
+        -P fuzz
+    exit_code=$?
+    delete_table "${BIGTABLE_PROJECT_ID}" "${BIGTABLE_INSTANCE_ID}" "${TABLE_ID}"
+    return $exit_code
+}
+
+run_pyspark_test() {
+    SPARK_VERSION=$1
+    HADOOP_VERSION=$2
+    # TODO: Install Python 3.7 on Kokoro machines to enable PySpark 2.4.8 tests
+    #   without running into the "TypeError: an integer is required" issue.
+    if [[ ${SPARK_VERSION} == "2.4.8" ]]; then return; fi
+
+    echo "***Running the PySpark test for Spark ${SPARK_VERSION}.***"
+    SPARK_BIN_NAME="spark-${SPARK_VERSION}-bin-hadoop${HADOOP_VERSION}"
+    BIGTABLE_SPARK_JAR=$(get_bigtable_spark_jar)
+    PYSPARK_TEST_SCRIPT="spark-bigtable_2.12/test-pyspark/read_and_write_test.py"
+    BASE_SCRIPT="spark-bigtable_2.12/test-pyspark/test_base.py"
+    # Download Apache Spark on the machine from a GCS bucket, new versions
+    # must be manually uploaded to the bucket first.
+    gsutil -q cp "gs://bigtable-spark-test-resources/spark-files/${SPARK_BIN_NAME}.tgz" .
+    tar xzf ${SPARK_BIN_NAME}.tgz
+    PYSPARK_TABLE_ID=$(create_table_id "pyspark")
+    ./${SPARK_BIN_NAME}/bin/spark-submit \
+        --packages=org.slf4j:slf4j-reload4j:1.7.36 \
+        --jars ${BIGTABLE_SPARK_JAR} \
+        --py-files=${BASE_SCRIPT} \
+        ${PYSPARK_TEST_SCRIPT} \
+        --bigtableProjectId=${BIGTABLE_PROJECT_ID} \
+        --bigtableInstanceId=${BIGTABLE_INSTANCE_ID} \
+        --bigtableTableId=${PYSPARK_TABLE_ID}
+    exit_code=$?
+    delete_table ${BIGTABLE_PROJECT_ID} ${BIGTABLE_INSTANCE_ID} ${PYSPARK_TABLE_ID}
+    return $exit_code
+}
+
+case ${JOB_TYPE} in
+presubmit)
+    run_bigtable_spark_tests "3.1.3" "integration"
+    RETURN_CODE=$?
+    run_pyspark_test "3.1.3" "3.2"
+    RETURN_CODE=$(($RETURN_CODE || $?))
+    run_unit_tests
+    RETURN_CODE=$(($RETURN_CODE || $?))
+    ;;
+all_versions)
```

Comment:
Wont this run them serially? Wouldnt it be better to have them run in parallel jobs?

Reply:
I agree, but `all_versions` is run periodically every 4 hours in the background, so not sure how much difference making it parallel and faster to finish makes.